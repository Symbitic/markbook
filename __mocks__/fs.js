/* global jest */

const fs = jest.requireActual('fs')

let contents = {}

const writeFile = (filename, data, callback) => {
  contents[filename] = /\ufffd/.test(data.toString()) ? data : data.toString()
  callback(undefined, data)
}

const filesystem = () => contents

const clear = () => {
  contents = {}
}

module.exports = {
  ...fs,
  writeFile,
  filesystem,
  clear
}
