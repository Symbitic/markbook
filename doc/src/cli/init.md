# `init`

The init command generates a new book:

    markbook init [dir]

## Directory

By default, markbook will use in the current directory. To have markbook use a different directory, simply give it as a commandline argument.

    markbook build path/to/book

## Options

| option | description | required? |
| ------ | ----------- | --------- |
| `--author` | Author's name | Yes |
| `--description` | Book description | No |
| `--title` | Book title | Yes |

If any of these three options are not given on the command-line, Markbook will prompt for them.

Non-required options can be empty.

