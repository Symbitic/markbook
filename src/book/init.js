import { createPath, readdir, readFile, writeFile } from '../common/files'
import path from 'path'
import { reject } from '../common/errors'
import { status } from '../common/log'

const copy = (author, desc, theme, title, origin, dir, file) => {
  const replace = data =>
    data
      .toString()
      .replace(/@TITLE@/g, title)
      .replace(/@DESC@/g, desc)
      .replace(/@AUTHOR@/g, author)
      .replace(/@THEME@/g, theme ? 'theme: "theme"' : '')
      .replace(/\n\n/g, '\n')

  const name = path.relative(origin, file)
  const output = path.join(dir, name)

  return readFile(file).then(data => writeFile(output, replace(data)))
}

/**
 * Create a new book.
 * @param {!string} dir Path to create a new book in.
 */
export default function init (dir, options) {
  const { author, desc, theme, title } = options

  if (!author) {
    return reject('Missing "author" field')
  } else if (!desc) {
    return reject('Missing "desc" field')
  } else if (!title) {
    return reject('Missing "title" field')
  }

  const defaultDir = createPath('default')
  const themeDir = createPath('theme')

  const args = [author, desc, theme, title]

  const copyFiles = () =>
    readdir(defaultDir).then(files =>
      Promise.all(files.map(file => copy(...args, defaultDir, dir, file)))
    )

  const copyTheme = () =>
    theme
      ? readdir(themeDir).then(files =>
          Promise.all(
            files.map(file =>
              copy(...args, themeDir, path.join(dir, 'theme'), file)
            )
          )
        )
      : Promise.resolve()

  return Promise.all([copyFiles(), copyTheme()]).then(() => status('Finished'))
}
