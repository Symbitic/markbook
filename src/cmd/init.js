import { status } from 'common/log'

/**
 * Create a new markbook.
 */
export default function (name, options = {}) {
  options.theme = options.theme || 'default'
  if (name) {
    status(`Init in ${name}`)
  } else {
    status('Init in current directory')
  }
  status(`Theme: ${options.theme}`)
}
