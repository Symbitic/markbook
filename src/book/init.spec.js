/* eslint-env jest */
import fs from 'fs'
import init from './init'

jest.mock('fs')

afterEach(fs.clear)

const case1 = [
  'create a new book',
  'book-1',
  'The Israelites',
  'The (nonliteral) word of God',
  'The Bible'
]

const case2 = [
  'reject without an author',
  'bad-author',
  '',
  'A book without an author',
  'Author-less book'
]

const case3 = [
  'reject without a description',
  'bad-desc',
  'Nobody',
  '',
  'Description-less book'
]

const case4 = [
  'reject without a title',
  'bad-title',
  'Nobody',
  'A book without a name',
  ''
]

const tests = [case1, case2, case3, case4]

const snapshot = (book, author, desc, title) =>
  expect(
    init(book, { author, desc, title }).then(fs.filesystem)
  ).resolves.toMatchSnapshot()

const missingAuthor = (book, author, desc, title) =>
  expect(
    init(book, { author, desc, title }).then(fs.filesystem)
  ).rejects.toThrow('Missing "author" field')

const missingDesc = (book, author, desc, title) =>
  expect(
    init(book, { author, desc, title }).then(fs.filesystem)
  ).rejects.toThrow('Missing "desc" field')

const missingTitle = (book, author, desc, title) =>
  expect(
    init(book, { author, desc, title }).then(fs.filesystem)
  ).rejects.toThrow('Missing "title" field')

describe.each(tests)('book/init', (name, book, author, desc, title) => {
  it(`should ${name}`, () => {
    if (!author) {
      return missingAuthor(book, author, desc, title)
    } else if (!desc) {
      return missingDesc(book, author, desc, title)
    } else if (!title) {
      return missingTitle(book, author, desc, title)
    } else {
      return snapshot(book, author, desc, title)
    }
  })
})
