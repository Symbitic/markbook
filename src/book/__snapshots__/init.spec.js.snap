// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`book/init should create a new book 1`] = `
Object {
  "book-1/markbook.yml": "title: The Bible
description: \\"The (nonliteral) word of God\\"
authors: [ 'The Israelites' ]
",
  "book-1/src/README.md": "# The Bible
This is your book.
",
  "book-1/src/SUMMARY.md": "# Summary
[The Bible](README.md)
* [Chapter 1](chapter-1.md)
",
  "book-1/src/chapter-1.md": "# Chapter 1
This is the first chapter.
",
}
`;
